admin/BundleSolrBackup

```diff
@@ -0,0 +1,236 @@
+later version: http://www.gnu.org/licenses/gpl-2.0.txt
+
+=cut
+
+################################################################################
+
+my $working_dir;
+my $debug_flag;
+my $help_flag;
+
+GetOptions(
+    'working-dir|w=s'            => \$working_dir,
+    'debug|d'                   => \$debug_flag,
+    'help|h'                    => \$help_flag,
+);
+
+pod2usage() if $help_flag;
+pod2usage(
+    -exitval => 64, # EX_USAGE
+    -message => "$0: unrecognized arguments",
+) if @ARGV;
+
+################################################################################
+
+$SIG{'INT'} = sub { exit 3 };
+
+use File::Copy qw( copy move );
+use POSIX qw( strftime );
+use String::ShellQuote qw( shell_quote );
+
+use FindBin;
+use lib "$FindBin::Bin/../lib";
+
+use DBDefs;
+use MusicBrainz::Script::MBDump qw( gpg_sign );
+
+################################################################################
+
+sub _debug { print '[debug] ', (sprintf shift, @_), "\n" if $debug_flag; }
+sub _info { print '[info] ', (sprintf shift, @_), "\n"; }
+sub _warn { warn '[warning] ', (sprintf shift, @_), "\n"; }
+sub _error { warn '[error] ', (sprintf shift, @_), "\n"; }
+
+my $BACKUP_STAMP = $ENV{BACKUP_STAMP};
+my $COMPRESSION_LEVEL = DBDefs->SOLR_BACKUP_COMPRESSION_LEVEL;
+my $COMPRESSION_THREADS = DBDefs->DUMP_COMPRESSION_THREADS;
+
+my @backup_dirs;
+my @extraneous_files;
+
+opendir(my $wdh, $working_dir) or die $OS_ERROR;
+for my $filename (sort grep { ! /^\.\.?$/ } readdir $wdh)
+{
+    if ($filename =~ /^backup_${BACKUP_STAMP}_/)
+    {
+        push @backup_dirs, $filename;
+    }
+    else
+    {
+        push @extraneous_files, $filename;
+    }
+}
+closedir $wdh;
+
+die "$working_dir has extraneous files: @extraneous_files\n"
+    if scalar @extraneous_files;
+
+my @public_domain_collections = qw(
+    area
+    artist
+    event
+    instrument
+    label
+    place
+    recording
+    release
+    release-group
+    series
+    url
+    work
+);
+
+my @private_collections = qw(
+    editor
+);
+
+for my $backup_dir (@backup_dirs)
+{
+    my $collection = $backup_dir =~ s/^backup_\d{8}-\d{6}_//r;
+    my $dump_dir = $collection;
+    opendir(my $bdh, "$working_dir/$backup_dir") or die $OS_ERROR;
+    if (any { /^$collection$/ } @private_collections)
+    {
+        system 'rm', '-fr', "$working_dir/$backup_dir";
+        next;
+    }
+    @extraneous_files = ();
+    for my $filename (sort grep { ! /^\.\.?$/ } readdir $bdh)
+    {
+        push @extraneous_files, $filename
+            if $filename ne $collection;
+    }
+    closedir $bdh;
+    die "$working_dir/$backup_dir has extraneous files: @extraneous_files\n"
+        if scalar @extraneous_files;
+    opendir(my $bsdh, "$working_dir/$backup_dir/$collection") or die $OS_ERROR;
+    for my $filename (sort grep { ! /^\.\.?$/ } readdir $bsdh)
+    {
+        push @extraneous_files, $filename
+            if ! $filename =~ 'backup_0\.properties|index|shard_backup_metadata|zk_backup';
```

Comment:
I think the operator precedence might be wrong here: the `!` will be applied to `$filename` only. I guess `!~` should be used instead.

Reply:
Correct. Now fixed through the commit https://github.com/metabrainz/musicbrainz-server/pull/3525/commits/dd2406f9fe8f1f1857f0206eeec0ba24bf7caceb.

Reply:
> I did review the script and found a possible syntax issue; not sure if that should have come up during testing or not.

For the record, this comparison code came from the old script I wrote some years ago. Its purpose was to prevent unexpected files to be archived and published. It was likely not preventing anything.